### PR TITLE
Add MANIFEST.in

### LINKED ISSUE #52843: setup.py sdist does not include third party submodules
## 🐛 Bug

`python setup.py sdist` should produce a tarball of source code that is self-contained and sufficient to compile the project. Currently
- the third-party sources are not included
- the LICENSE file does not include the third party licenses (solved for wheels by gh-51634)
- some other files that are required for building like `version.txt` are not included

This is related to issues gh-50697 and gh-13586

@rgommers

cc @malfet @seemethere @walterddr

### PR DESCRIPTION
**Pull request by malfet**
Do not build PyTorch if `setup.py` is called with  'sdist' option
Regenerate bundled license while sdist package is being built
Refactor `check_submodules` out of `build_deps` and check that submodules project are present during source package build stage.

Test that sdist package is configurable during `asan-build` step

Fixes #52843

### CODE REVIEW

**Comment by mattip** [comment]
In order to correctly concatentate the two license files `./LICENSE` and `./third_party/LICENSES_BUNDLED.txt`, the `sdist` command should be wrapped like the `class wheel_concatenate(bdist_wheel)` class.

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+recursive-include test *.*
+recursive-include docs *.*
+recursive-include ios *.*
+recursive-include third_party *
+recursive-include test *.*
+recursive-include benchmarks *.*
+recursive-include scripts *.*
+recursive-include mypy_plugins *.*
+recursive-include modules *.*
+global-exclude *.o *.so *.dylib *.a .git
```
Comment (rgommers) [inline MANIFEST.in:19]: You'll probably want to ensure to not do this in a dirty tree.

E.g. what SciPy does: https://github.com/scipy/scipy/blob/master/setup.py#L202

- in the `sdist` command in `setup.py`, check if all submodules are up to date.
- In the release script, ensure a `git clean -xdf` or running it on a clean git clone.

Comment (xuzhao9) [inline MANIFEST.in:19]: Maybe we should add more global-exclude file patterns similar to SciPy, such as `*.swp`, `*.pyc`? https://github.com/scipy/scipy/blob/master/MANIFEST.in#L31

Comment (malfet) [reply MANIFEST.in:19]: Agree about the clean, but not necessarily about keeping submodules uptodate, as one should be at free to package their repo in whatever state of dirtiness into their `sdist` and `bdist` packages.

Comment (rgommers) [reply MANIFEST.in:19]: You can always just commit whatever you want to package on a branch (even a temporary one), right? I'm not sure what the use case is for packaging things that aren't  committed anywhere.

Up to you of course, I just know SciPy implemented it after shipping an incomplete release and we've never had a complaint about the check.

Comment (malfet) [reply MANIFEST.in]: I've extracted existing submodules validation logic and share it across 'sdist' and 'build_deps' steps, will add some of your suggestions in the followup up diff.

**Review by walterddr — APPROVED** [review]

**Review by rgommers — APPROVED** [review]

**Review by seemethere — APPROVED** [review]

File: .jenkins/pytorch/build.sh [hunk 2]
```diff
@@ -317,3 +317,15 @@ if [[ "${BUILD_ENVIRONMENT}" == *xla* ]]; then
   popd
   assert_git_not_dirty
 fi
+
+# Test that source package is configurable
```
Comment (mattip) [inline .jenkins/pytorch/build.sh:321]:
```suggestion
# Test building via the sdist source tarball
```

**Comment by mattip** [comment]
I expected to see some traces of the build in the [pytorch_linux_xenial_py3_clang5_asan_build](https://app.circleci.com/pipelines/github/pytorch/pytorch/278769/workflows/3a9cc440-3df6-4e72-90ee-4e8e09af21d0/jobs/11193255/parallel-runs/0/steps/0-106), but it seems that job calls [`exec .. build-asan.sh`](https://github.com/pytorch/pytorch/blob/v1.8.0-rc1/.jenkins/pytorch/build.sh#L15) and exits without continuing. Is there another `*asan*` build that hits the new test?

File: setup.py [hunk 3]
```diff
@@ -623,8 +628,14 @@ def run(self):
                         except OSError:
                             shutil.rmtree(filename, ignore_errors=True)
 
-        # It's an old-style class in Python 2.7...
-        distutils.command.clean.clean.run(self)
+        super().run(self)
```
Comment (malfet) [inline setup.py:631]: Fixed typo
```suggestion
        super().run()
```

**Comment by malfet** [comment]
> I expected to see some traces of the build in the [pytorch_linux_xenial_py3_clang5_asan_build](https://app.circleci.com/pipelines/github/pytorch/pytorch/278769/workflows/3a9cc440-3df6-4e72-90ee-4e8e09af21d0/jobs/11193255/parallel-runs/0/steps/0-106), but it seems that job calls [`exec .. build-asan.sh`](https://github.com/pytorch/pytorch/blob/v1.8.0-rc1/.jenkins/pytorch/build.sh#L15) and exits without continuing. Is there another `*asan*` build that hits the new test?

Thank you for spotting. Moved the test logic from `build.sh` to `build-asan.sh`

**Comment by mattip** [comment]
Cool. 
The [build log](https://app.circleci.com/pipelines/github/pytorch/pytorch/279654/workflows/e80fd11f-1ad1-46ed-a8ed-49cf929b84ac/jobs/11235905) has a `warning: sdist: standard file not found: should have one of README, README.txt` at the beginning of `python setup.py sdist`, and lots and lots of lines about creating the sdist. 

Then at the beginning of the cmake-only build (`python setup.py build --cmake-only`) there is a `fatal: Not a git repository (or any of the parent directories): .git` but it is not really fatal. Other than that, the cmake warnings seem to be the usual ones I see when building from a git checkout.

**Pull request closed without merging**